Digit Match

shrutivanw opened the pull request:
CS Fundamentals' assignment to author a method that returns the count of matching digits in the two input non-negative integers.

shrutivanw committed:
Digit Match

CS Fundamentals' assignment to author a method that returns the count of matching digits in the two input non-negative integers.

shrutivanw committed:
Digit Match test update

Adding test for nil

shrutivanw committed:
Additional tests

Fixing tests for nil and adding tests for negative input.

Hamled commented on README.md:
```diff
@@ -1 +1,15 @@
-# digit_match
\ No newline at end of file
+# Digit Match
+Number of matching digits starting with the least significant digit in two integers.
```
This is sort of like the opposite of hamming distance (if the shorter number is zero-padded to the same length), is there any value in mentioning it to them or otherwise describing the problem in terms of that idea?

The students are given a weekend warrior exercise to implement a hamming distance algorithm, so it's not outside of their experience depending on when this gets assigned.

shrutivanw replied:
That's a great context to mention! So as not to increase the homework by adding required reading, I added a mention at the bottom of readme file as an aside.

Hamled commented on README.md:
```diff
@@ -1 +1,15 @@
+The method should return 4 in this case because 4 of these pairs match (2-2, 5-5, 8-8, and 1-1).
+* Share and explain the time and space complexities for your solution.
+* If you think of multiple approaches to solve the problem, implement the solution which minimizes space complexity. Explain the other approaches, and your decision in comments above the code.
+**Note**: Do not use Ruby provided functionality.
```
We should have an extra blank line above this so that the note goes into its own paragraph.

shrutivanw replied:
Good catch! Now corrected.

Hamled commented on lib/digit_match.rb:
```diff
@@ -0,0 +1,5 @@
+# Returns count of digits matching in the two input non-negative integers
+def digit_match(number_1, number_2)
+  puts "NOT IMPLEMENTED"
+  return -1
```
How would you feel about doing `raise NotImplementedError` here instead?

Best practice I think is generally to avoid `puts` statements and other things that will print to the console for code being tested, because that output kind of gets randomly interspersed with the test runner's output.

Hamled replied:
Also this is a very minor quibble, but something irks me about the variable names `number_1` and `number_2`. Perhaps we could do `number1` and `number2` or even `left` and `right`.

shrutivanw replied:
Both now corrected.

Hamled commented on specs/digit_match_spec.rb:
```diff
@@ -0,0 +1,61 @@
+
+describe "digit match" do
+  describe "basic tests" do
+    describe "test 1" do
```
Same as the other repos, we should probably remove the `describe` blocks that are immediate parents of individual tests.

Hamled commented on specs/digit_match_spec.rb:
```diff
@@ -0,0 +1,61 @@
+  # if the parameter is an object, check for nil
+  describe "nil object input test" do
+    describe "test 1" do
+      it "first input is nil" do
```
I could see this and the other nil test going into the edge cases section.

shrutivanw committed:
Incorporating CR feedback

- Changing spec to raise error
- Adding link for further reading for Hamming distance problem and cleaning up the readme file to render correctly
- Cleaning up the spec to be consistent with other specs (nil tests inside edge cases, removing nested describe inside basic and edge case tests)

shrutivanw commented:
Thank you @Hamled Feedback addressed.
